File: cmd/update.go
package cmd

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

const (
	githubRepo     = "dalang-io/dalang-cli"
	updateBaseURL  = "https://github.com/" + githubRepo + "/releases/latest/download"
	versionInfoURL = "https://api.github.com/repos/" + githubRepo + "/releases/latest"
)

var updateHTTPClient = &http.Client{Timeout: 60 * time.Second}

type VersionInfo struct {
	Version string `json:"tag_name"`
}

func cmdUpdate(args []string) error {
	if len(args) > 0 && (args[0] == "--help" || args[0] == "-h") {
		printUpdateHelp()
		return nil
	}

	printInfo("Checking for updates...")

	// Get current version
	currentVersion := Version

	// Check latest version
	latestVersion, err := getLatestVersion()
	if err != nil {
		printWarn("Could not check for updates: %v", err)
		printInfo("Proceeding with update anyway...")
		latestVersion = &VersionInfo{Version: "latest"}
	} else {
		if latestVersion.Version == currentVersion {
			printSuccess("Already up to date (version %s)", currentVersion)
			return nil
		}
		printInfo("New version available: %s (current: %s)", latestVersion.Version, currentVersion)
	}

	// Determine binary name based on OS/arch
	binaryName := getBinaryName()
	downloadURL := fmt.Sprintf("%s/%s", updateBaseURL, binaryName)

	printInfo("Downloading %s...", binaryName)

	// Get current executable path
	execPath, err := os.Executable()
	if err != nil {
		return fmt.Errorf("failed to get executable path: %w", err)
	}
	execPath, err = filepath.EvalSymlinks(execPath)
	if err != nil {
		return fmt.Errorf("failed to resolve executable path: %w", err)
	}

	// Download to temp file
	tmpFile, err := os.CreateTemp(filepath.Dir(execPath), ".dalang-update-*")
	if err != nil {
		return fmt.Errorf("failed to create temp file: %w", err)
	}
	tmpPath := tmpFile.Name()
	defer os.Remove(tmpPath)

	resp, err := updateHTTPClient.Get(downloadURL)
	if err != nil {
		return fmt.Errorf("failed to download: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("download failed: HTTP %d", resp.StatusCode)
	}

	hash := sha256.New()
	_, err = io.Copy(io.MultiWriter(tmpFile, hash), resp.Body)
	tmpFile.Close()
	if err != nil {
		return fmt.Errorf("failed to save download: %w", err)
	}

	// Verify checksum against checksums.txt published with the release
	checksumURL := updateBaseURL + "/checksums.txt"
	checksumResp, err := updateHTTPClient.Get(checksumURL)
	if err != nil {
		return fmt.Errorf("failed to fetch checksum: %w", err)
	}
	defer checksumResp.Body.Close()
	if checksumResp.StatusCode != http.StatusOK {
		return fmt.Errorf("checksum unavailable: HTTP %d", checksumResp.StatusCode)
	}
	checksumBody, err := io.ReadAll(checksumResp.Body)
	if err != nil {
		return fmt.Errorf("failed to read checksum: %w", err)
	}
	expectedHex := ""
	for _, line := range strings.Split(string(checksumBody), "\n") {
		fields := strings.Fields(strings.TrimSpace(line))
		if len(fields) >= 2 && fields[len(fields)-1] == binaryName {
			expectedHex = fields[0]
			break
		}
	}
	if expectedHex == "" {
		return fmt.Errorf("checksum for %s not found in checksums.txt", binaryName)
	}
	actualHex := hex.EncodeToString(hash.Sum(nil))
	if !strings.EqualFold(expectedHex, actualHex) {
		return fmt.Errorf("checksum mismatch: expected %s, got %s — download may be corrupted", expectedHex, actualHex)
	}
	PrintDebug("Checksum verified: %s", actualHex)

	// Make executable
	if err := os.Chmod(tmpPath, 0755); err != nil {
		return fmt.Errorf("failed to set permissions: %w", err)
	}

	// Replace current binary
	// On Windows, we need to rename differently
	if runtime.GOOS == "windows" {
		oldPath := execPath + ".old"
		os.Remove(oldPath) // Remove any existing .old file
		if err := os.Rename(execPath, oldPath); err != nil {
			return fmt.Errorf("failed to backup old binary: %w", err)
		}
		if err := os.Rename(tmpPath, execPath); err != nil {
			// Try to restore
			os.Rename(oldPath, execPath)
			return fmt.Errorf("failed to install new binary: %w", err)
		}
		os.Remove(oldPath)
	} else {
		// Unix: atomic rename
		if err := os.Rename(tmpPath, execPath); err != nil {
			// May need sudo - try to give helpful message
			if os.IsPermission(err) {
				printError("Permission denied. Try: sudo dalang update")
				return fmt.Errorf("permission denied")
			}
			return fmt.Errorf("failed to install new binary: %w", err)
		}
	}

	if latestVersion.Version != "latest" {
		printSuccess("Updated to version %s", latestVersion.Version)
	} else {
		printSuccess("Updated to latest version")
	}

	return nil
}

func getLatestVersion() (*VersionInfo, error) {
	resp, err := updateHTTPClient.Get(versionInfoURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("HTTP %d", resp.StatusCode)
	}

	var info VersionInfo
	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
		return nil, err
	}
	if info.Version == "" {
		return nil, fmt.Errorf("latest release tag not found")
	}

	return &info, nil
}

func getBinaryName() string {
	goos := runtime.GOOS
	arch := runtime.GOARCH

	ext := ""
	if goos == "windows" {
		ext = ".exe"
	}

	return fmt.Sprintf("dalang-%s-%s%s", goos, arch, ext)
}

func printUpdateHelp() {
	fmt.Printf(`%sdalang update%s - Update CLI to latest version

%sUSAGE:%s
    dalang update

%sDESCRIPTION:%s
    Downloads and installs the latest version of Dalang CLI.
    On Linux/macOS, may require sudo if installed in /usr/local/bin.

%sEXAMPLES:%s
    dalang update              # Update to latest version
    sudo dalang update         # Update with elevated permissions

%sNOTE:%s
    Current binary will be replaced in-place.
    Your credentials and config are preserved.
`,
		colorCyan, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
		colorYellow, colorReset,
	)
}

File: src/commands/update.rs
use std::env;
use std::fs;
use std::io::Read;

use crate::api;
use crate::output::*;

const UPDATE_BASE_URL: &str = "https://dalang.io/cli";
const VERSION_INFO_URL: &str = "https://dalang.io/cli/version.json";

pub fn cmd_update(args: &[String]) -> Result<(), String> {
    if !args.is_empty() && (args[0] == "--help" || args[0] == "-h") {
        crate::commands::help::print_update_help();
        return Ok(());
    }

    print_info("Checking for updates...");

    let current_version = crate::commands::version::get_version();

    let latest_version = match get_latest_version() {
        Ok(info) => {
            if info.version == current_version {
                print_success(&format!("Already up to date (version {})", current_version));
                return Ok(());
            }
            print_info(&format!(
                "New version available: {} (current: {})",
                info.version, current_version
            ));
            Some(info)
        }
        Err(e) => {
            print_warn(&format!("Could not check for updates: {}", e));
            print_info("Proceeding with update anyway...");
            None
        }
    };

    let binary_name = get_binary_name();
    let download_url = format!("{}/{}", UPDATE_BASE_URL, binary_name);

    print_info(&format!("Downloading {}...", binary_name));

    // Get current executable path
    let exec_path = env::current_exe()
        .map_err(|e| format!("failed to get executable path: {}", e))?;
    let exec_path = fs::canonicalize(&exec_path)
        .map_err(|e| format!("failed to resolve executable path: {}", e))?;

    // Download to temp file
    let tmp_dir = env::temp_dir();
    let tmp_path = tmp_dir.join(format!("dalang-update-{}", std::process::id()));

    let tls = native_tls::TlsConnector::new()
        .map_err(|e| format!("TLS init error: {}", e))?;
    let agent = ureq::AgentBuilder::new()
        .timeout(std::time::Duration::from_secs(120))
        .tls_connector(std::sync::Arc::new(tls))
        .resolver(crate::api::dns::CloudflareDns)
        .build();

    let resp = agent
        .get(&download_url)
        .call()
        .map_err(|e| format!("failed to download: {}", e))?;

    if resp.status() != 200 {
        return Err(format!("download failed: HTTP {}", resp.status()));
    }

    let mut body = Vec::new();
    resp.into_reader()
        .read_to_end(&mut body)
        .map_err(|e| format!("failed to save download: {}", e))?;

    fs::write(&tmp_path, &body).map_err(|e| format!("failed to write temp file: {}", e))?;

    // Make executable (Unix)
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        fs::set_permissions(&tmp_path, fs::Permissions::from_mode(0o755))
            .map_err(|e| format!("failed to set permissions: {}", e))?;
    }

    // Replace current binary
    if cfg!(windows) {
        let old_path = exec_path.with_extension("old");
        let _ = fs::remove_file(&old_path);
        fs::rename(&exec_path, &old_path)
            .map_err(|e| format!("failed to backup old binary: {}", e))?;
        if let Err(e) = fs::rename(&tmp_path, &exec_path) {
            let _ = fs::rename(&old_path, &exec_path);
            return Err(format!("failed to install new binary: {}", e));
        }
        let _ = fs::remove_file(&old_path);
    } else {
        if let Err(e) = fs::rename(&tmp_path, &exec_path) {
            if e.kind() == std::io::ErrorKind::PermissionDenied {
                let _ = fs::remove_file(&tmp_path);
                print_error("Permission denied. Try: sudo dalang update");
                return Err("permission denied".to_string());
            }
            let _ = fs::remove_file(&tmp_path);
            return Err(format!("failed to install new binary: {}", e));
        }
    }

    if let Some(info) = latest_version {
        print_success(&format!("Updated to version {}", info.version));
    } else {
        print_success("Updated to latest version");
    }

    Ok(())
}

fn get_latest_version() -> Result<api::VersionInfo, String> {
    let tls = native_tls::TlsConnector::new()
        .map_err(|e| format!("TLS init error: {}", e))?;
    let agent = ureq::AgentBuilder::new()
        .timeout(std::time::Duration::from_secs(10))
        .tls_connector(std::sync::Arc::new(tls))
        .resolver(crate::api::dns::CloudflareDns)
        .build();

    let resp = agent
        .get(VERSION_INFO_URL)
        .call()
        .map_err(|e| format!("{}", e))?;

    if resp.status() != 200 {
        return Err(format!("HTTP {}", resp.status()));
    }

    let mut body = Vec::new();
    resp.into_reader()
        .read_to_end(&mut body)
        .map_err(|e| format!("{}", e))?;

    serde_json::from_slice(&body).map_err(|e| format!("{}", e))
}

fn get_binary_name() -> String {
    let os = if cfg!(target_os = "windows") {
        "windows"
    } else if cfg!(target_os = "macos") {
        "darwin"
    } else if cfg!(target_os = "android") {
        "android"
    } else {
        "linux"
    };

    let arch = if cfg!(target_arch = "x86_64") {
        "amd64"
    } else if cfg!(target_arch = "aarch64") {
        "arm64"
    } else {
        "amd64"
    };

    let ext = if cfg!(target_os = "windows") {
        ".exe"
    } else {
        ""
    };

    format!("dalang-{}-{}{}", os, arch, ext)
}
